src/api-service/__app__/onefuzzlib/request_access.py

```diff
@@ -0,0 +1,115 @@
+from typing import Dict, List
+from uuid import UUID
+
+from onefuzztypes.models import ApiAccessRule
+from pydantic import BaseModel, parse_raw_as
+
+
+class RuleConflictError(Exception):
+    def __init__(self, message: str) -> None:
+        super(RuleConflictError, self).__init__(message)
+        self.message = message
+
+
+class RequestAccess:
+    """
+    Stores the rules associated with a the request paths
+    """
+
+    class Rules:
+        allowed_groups_ids: List[UUID]
+
+        def __init__(self, allowed_groups_ids: List[UUID] = []) -> None:
+            self.allowed_groups_ids = allowed_groups_ids
+
+    class Node:
+        # http method -> rules
+        rules: Dict[str, "RequestAccess.Rules"]
+        # path -> node
+        children: Dict[str, "RequestAccess.Node"]
+
+        def __init__(self) -> None:
+            self.rules = {}
+            self.children = {}
+            pass
+
+    root: Node
+
+    def __init__(self) -> None:
+        self.root = RequestAccess.Node()
+
+    def __add_url__(self, methods: List[str], path: str, rules: Rules) -> None:
+        methods = list(map(lambda m: m.upper(), methods))
+
+        segments = path.split("/")
+        if len(segments) == 0:
```

Comment:
if length is 0, does it mean that we apply rules to all APIs ?

Reply:
no it mens there are no rules